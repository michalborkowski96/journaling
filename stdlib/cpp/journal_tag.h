template<typename T>
class Type_JournalTag_;

template<typename T>
class Type_JournalTag {
public:
	🍆::WeakObject<Type_JournalTag> this_ptr;
	🍆::StrongObject<Type_JournalTag> get_this() {
		return 🍆::scast<Type_JournalTag>(this_ptr);
	}
	🍆::StrongObject<T> object;
	🍆::Journal::Tag tag;
	virtual ~Type_JournalTag() = default;

	Type_JournalTag(WeakObject<Type_JournalTag> this_ptr, 🍆::StrongObject<T> var_t) : this_ptr(this_ptr), object(var_t), tag(var_t->journal.get_tag()){
	}

	Type_JournalTag() = delete;

	Type_JournalTag(const Type_JournalTag&) = delete;

	Type_JournalTag(Type_JournalTag&&) = delete;

	void basefun_bring_back() {
		object->journal.goto_tag(tag);
	}

	virtual void privfun_bring_back() {
		(basefun_bring_back());
	}
	virtual void fun_bring_back() {
		(basefun_bring_back());
	}
	void basefun_t_(🍆::StrongObject<Type_JournalTag_<T>> var_ttt, 🍆::StrongObject<T> var_tt) {
		((var_ttt)->fun_t((var_tt)));
	}

	virtual void privfun_t_(🍆::StrongObject<Type_JournalTag_<T>> var_ttt, 🍆::StrongObject<T> var_tt) {
		(basefun_t_(var_ttt, var_tt));
	}
	virtual void fun_t_(🍆::StrongObject<Type_JournalTag_<T>> var_ttt, 🍆::StrongObject<T> var_tt) {
		(basefun_t_(var_ttt, var_tt));
	}
};

template<>
class Type_JournalTag<🍆::Integer> {
public:
	🍆::WeakObject<Type_JournalTag> this_ptr;
	🍆::StrongObject<Type_JournalTag> get_this() {
		return 🍆::scast<Type_JournalTag>(this_ptr);
	}
	virtual ~Type_JournalTag() = default;

	Type_JournalTag(WeakObject<Type_JournalTag> this_ptr, 🍆::StrongObject<🍆::Integer>) : this_ptr(this_ptr) {
	}

	Type_JournalTag() = delete;

	Type_JournalTag(const Type_JournalTag&) = delete;

	Type_JournalTag(Type_JournalTag&&) = delete;

	void basefun_bring_back() {}

	virtual void privfun_bring_back() {
		(basefun_bring_back());
	}
	virtual void fun_bring_back() {
		(basefun_bring_back());
	}
	void basefun_t_(🍆::StrongObject<Type_JournalTag_<🍆::Integer>>, 🍆::StrongObject<🍆::Integer>) {}

	virtual void privfun_t_(🍆::StrongObject<Type_JournalTag_<🍆::Integer>> var_ttt, 🍆::StrongObject<🍆::Integer> var_tt) {
		(basefun_t_(var_ttt, var_tt));
	}
	virtual void fun_t_(🍆::StrongObject<Type_JournalTag_<🍆::Integer>> var_ttt, 🍆::StrongObject<🍆::Integer> var_tt) {
		(basefun_t_(var_ttt, var_tt));
	}
};

#define 🍆_Default_JournalTag
