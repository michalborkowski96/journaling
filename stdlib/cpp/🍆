#ifndef üçÜ
#define üçÜ

#include "tagged_tree.h"

#include <variant>
#include <functional>
#include <memory>
#include <type_traits>
#include <tuple>

namespace üçÜ {
	template <typename... Fs>
	struct overload : std::remove_reference_t<Fs>... {
		overload(Fs... fs) : std::remove_reference_t<Fs>{std::forward<Fs>(fs)}...{}
		using std::remove_reference_t<Fs>::operator()...;
	};

	class Null {};

	Null null;

	template<typename T>
	class StrongObject;

	class Integer;

	template<>
	class StrongObject<Integer> {
		std::variant<long long, std::shared_ptr<std::function<StrongObject<Integer>()>>> value;
	public:
		using object_type = Integer;
		StrongObject<Integer>();
		StrongObject<Integer>(const Null&);
		StrongObject<Integer>(Integer value);
		StrongObject<Integer>(long long value);
		StrongObject<Integer>(std::shared_ptr<std::function<StrongObject<Integer>()>> fun) : value(std::move(fun)) {}

		Integer& operator*();
		Integer* operator->();
		StrongObject<Integer> operator==(StrongObject<Integer> b);
		StrongObject<Integer> operator!=(StrongObject<Integer> b);

		explicit operator bool();

		long long& get_value() {
			return std::visit(overload(
			[](long long& v)->long long&{return v;},
			[&](std::shared_ptr<std::function<StrongObject<Integer>()>>& f)->long long&{
				long long val = (*f)().get_value();
				*f = [=](){
					return StrongObject<Integer>(val);
				};
				value = val;
				return std::get<long long>(value);
			}
			), value);
		}

		friend class Integer;
	};

	class Integer {
		long long value;
	public:
		Integer() = delete;
		StrongObject<Integer> fun_add(StrongObject<Integer> o) const {
			return StrongObject<Integer>(value + o.get_value());
		}
		StrongObject<Integer> fun_sub(StrongObject<Integer> o) const {
			return StrongObject<Integer>(value - o.get_value());
		}
		StrongObject<Integer> fun_mul(StrongObject<Integer> o) const {
			return StrongObject<Integer>(value * o.get_value());
		}
		StrongObject<Integer> fun_div(StrongObject<Integer> o) const {
			return StrongObject<Integer>(value / o.get_value());
		}
		StrongObject<Integer> fun_mod(StrongObject<Integer> o) const {
			return StrongObject<Integer>(value % o.get_value());
		}
		StrongObject<Integer> fun_alt(StrongObject<Integer> o) const {
			return StrongObject<Integer>(value || o.get_value());
		}
		StrongObject<Integer> fun_less(StrongObject<Integer> o) const {
			return StrongObject<Integer>(value < o.get_value());
		}
		StrongObject<Integer> fun_le(StrongObject<Integer> o) const {
			return StrongObject<Integer>(value <= o.get_value());
		}
		StrongObject<Integer> fun_more(StrongObject<Integer> o) const {
			return StrongObject<Integer>(value > o.get_value());
		}
		StrongObject<Integer> fun_me(StrongObject<Integer> o) const {
			return StrongObject<Integer>(value >= o.get_value());
		}
		StrongObject<Integer> fun_opp() const {
			return StrongObject<Integer>(-value);
		}
		StrongObject<Integer> fun_neg() const {
			return StrongObject<Integer>(!value);
		}
		StrongObject<Integer> fun_con(StrongObject<Integer> o) const {
			return StrongObject<Integer>(value && o.get_value());
		}
/*
		StrongObject<Integer> operator==(StrongObject<Integer> b) const {
			return StrongObject<Integer>(value == b.get_value());
		}
		StrongObject<Integer> operator!=(StrongObject<Integer> b) const {
			return StrongObject<Integer>(value != b.get_value());
		}
*/
		static Integer* cast_from(long long* value) {
			return reinterpret_cast<Integer*>(value);
		}
/*
		explicit operator bool() const {
			return value;
		}
*/
		long long& to_ll() {
			return value;
		}
	};

	StrongObject<Integer>::StrongObject() : value(0LL) {}

	StrongObject<Integer>::StrongObject(const Null&) : value(0LL) {}

	StrongObject<Integer>::StrongObject(Integer value) : value(value.to_ll()) {}

	StrongObject<Integer>::StrongObject(long long value) : value(value) {}

	Integer& StrongObject<Integer>::operator*() {
		return *Integer::cast_from(&get_value());
	}
	Integer* StrongObject<Integer>::operator->() {
		return Integer::cast_from(&get_value());
	}
	StrongObject<Integer> StrongObject<Integer>::operator==(StrongObject<Integer> b) {
		return std::visit(overload(
		[&](long long& v){return v == b.get_value();},
		[&](std::shared_ptr<std::function<StrongObject<Integer>()>>& fa){
			return std::visit(overload(
			[&](long long& v){return v == get_value();},
			[&](std::shared_ptr<std::function<StrongObject<Integer>()>>& fb){
				if(fa == fb) {
					return true;
				}
				return get_value() == b.get_value();
			}
			), b.value);
		}
		), value);
	}
	StrongObject<Integer> StrongObject<Integer>::operator!=(StrongObject<Integer> b) {
		return !(*this == b);
	}

	StrongObject<Integer>::operator bool() {
		return get_value();
	}

	template<typename T>
	class WeakObject;

	const std::shared_ptr<std::function<std::shared_ptr<void>()>> null_ptr_fun = std::make_shared<std::function<std::shared_ptr<void>()>>([]()->std::shared_ptr<void>{return nullptr;});

	template<typename T>
	class StrongObject {
		std::shared_ptr<std::function<std::shared_ptr<void>()>> obj;
		friend class WeakObject<T>;
		template<typename U>
		friend class StrongObject;
	public:
		using object_type = T;
		T& operator*() {
			return *get_value();
		}
		T* operator->() {
			return get_value().get();
		}
		StrongObject(const Null&) : obj(null_ptr_fun) {}
		StrongObject() : obj(null_ptr_fun) {}
		StrongObject(const WeakObject<T>& o);
		StrongObject(std::shared_ptr<std::function<std::shared_ptr<void>()>> fun) : obj(std::move(fun)) {}
		template<typename U, std::enable_if_t<std::is_convertible<U*, T*>::value, int> = 0>
		StrongObject(StrongObject<U> o) : obj(o.obj) {}

		template<typename... Ts>
		static StrongObject make_object(Ts&&... args) {
			StrongObject ret;
			char* memory = new char[sizeof(T)];
			std::shared_ptr<T> ptr(reinterpret_cast<T*>(memory), std::function<void(T*)>([](T* ptr){
				delete[](reinterpret_cast<char*>(ptr));
			}));
			ret.obj = std::make_shared<std::function<std::shared_ptr<void>()>>([=]()->std::shared_ptr<void>{return ptr;});
			new (reinterpret_cast<T*>(memory)) T(ret, std::forward<Ts>(args)...);
			*std::get_deleter<std::function<void(T*)>>(ptr) = [](T* ptr){
				ptr->~T();
				delete[](reinterpret_cast<char*>(ptr));
			};
			return ret;
		}

		StrongObject<Integer> operator==(StrongObject<T> b) {
			if(obj == b.obj) {
				return true;
			}
			return get_value() == b.get_value();
		}
		StrongObject<Integer> operator!=(StrongObject<T> b) {
			return !(*this == b);
		}
/*
		explicit operator bool() const {
			return obj;
		}
*/
		std::shared_ptr<T> get_value() {
			std::shared_ptr<T> val = std::static_pointer_cast<T>((*obj)());
			*obj = [=](){
				return val;
			};
			return val;
		}

		template<typename U>
		static StrongObject scast(StrongObject<U> o) {
			return StrongObject(o.obj);
		}

		template<typename U>
		static StrongObject dcast(StrongObject<U> o) {
			std::shared_ptr<U> value = o.get_value();
			if(!value) {
				return null;
			}
			return std::make_shared<std::function<std::shared_ptr<void>()>>([val{std::dynamic_pointer_cast<T>(value)}](){return val;});
		}
	};

	template<typename T>
	class WeakObject {
		std::weak_ptr<std::function<std::shared_ptr<void>()>> obj;
		friend class StrongObject<T>;
		template<typename U>
		friend class WeakObject;
	public:
		using object_type = T;
		T& operator*() {
			return *get_value();
		}
		T* operator->() {
			return get_value().get();
		}
		WeakObject(const Null&) : obj(null_ptr_fun) {}
		WeakObject(std::weak_ptr<std::function<std::shared_ptr<T>()>> obj) : obj(obj) {}
		WeakObject() : obj(null_ptr_fun) {}
		WeakObject(const StrongObject<T>& o) : obj(o.obj) {}
		template<typename U, std::enable_if_t<std::is_convertible<U*, T*>::value, int> = 0>
		WeakObject(WeakObject<U> o) : obj(o.obj) {}
		StrongObject<Integer> operator==(WeakObject<T> b) {
			return StrongObject<T>(*this) == StrongObject<T>(b);
		}
		StrongObject<Integer> operator!=(WeakObject<T> b) {
			return !(*this == b);
		}
/*
		explicit operator bool() const {
			return obj;
		}
*/
		std::shared_ptr<T> get_value() {
			auto locked = obj.lock();
			std::shared_ptr<T> val = std::static_pointer_cast<T>((*locked)());
			*locked = [=](){
				return val;
			};
			return val;
		}
	};

	template<>
	class WeakObject<Integer>;

	template<typename T>
	StrongObject<T>::StrongObject(const WeakObject<T>& o) : obj(o.obj.lock()) {}

	template<typename T, typename... Ts>
	StrongObject<T> make_object(Ts&&... args) {
		return StrongObject<T>::make_object(std::forward<Ts>(args)...);
	}

	template<typename T>
	class has_journal;

	template<>
	class has_journal<Integer> : public std::false_type {};

	template<typename T, typename P, typename U = typename P::object_type>
	StrongObject<T> dcast(P o) {
		return StrongObject<T>::dcast(StrongObject<U>(o));
	}

	template<typename T, typename P, typename U = typename P::object_type>
	StrongObject<T> scast(P o) {
		return StrongObject<T>::scast(StrongObject<U>(o));
	}

	class Journal {
		template<typename...>
		struct unpack_to_args;

		template<typename R, typename O, typename... Ts>
		struct unpack_to_args<R, O, std::tuple<Ts...>> {
			using type = R (O::*)(Ts...);
		};

		template<typename...>
		struct unpack_to_args_skip_first;

		template<typename R, typename O, typename T, typename... Ts>
		struct unpack_to_args_skip_first<R, O, std::tuple<T, Ts...>> {
			using type = R (O::*)(Ts...);
		};

#warning TAG
		class Tag {};
		Tag get_tag() const;
		void force_tag(Tag);
		void goto_tag(Tag);

		template<typename...>
		struct count_types_with_journal;

		template<>
		struct count_types_with_journal<> {
			count_types_with_journal() = delete;
			static constexpr size_t value = 0;
		};

		template<typename T, typename... Ts>
		struct count_types_with_journal<T, Ts...> {
			count_types_with_journal() = delete;
			static constexpr size_t value = (has_journal<T>::value ? 1 : 0) + count_types_with_journal<Ts...>::value;
		};

#warning CAPTURED
		template<typename T, bool journal = has_journal<T>::value>
		class Captured;

		template<typename T>
		class Captured<T, true> {
			Tag tag;
			StrongObject<T> object;
			Captured() = delete;
		};

		template<typename T>
		class Captured<T, false> {
			StrongObject<T> object;
			Captured() = delete;
		};

		template<typename T, typename... Ts>
		auto trim_tuple(std::tuple<T, Ts...>&& tuple) {
			return std::apply([](const auto&, auto&&... args){return std::make_tuple(std::move(args)...);}, tuple);
		}

#warning CAPTURE
		template<typename... Args>
		static std::tuple<Captured<Args>...> capture(const StrongObject<Args>&... args);

#warning CAPTURED_GET
		template<typename... Args>
		static std::tuple<StrongObject<Args>...> captured_get(std::tuple<Captured<Args>...>& args);

#warning GET_TAGS
		template<typename... Args>
		static std::array<Tag, count_types_with_journal<Args...>::value> get_tags(const Captured<Args>&... args);

#warning GOTO_TAGS
		template<typename... Args>
		static void goto_tags(const std::array<Tag, count_types_with_journal<Args...>::value>& tags, std::tuple<Captured<Args>...>& args);

#warning ADD_NEW_VERTEX
		void add_new_vertex(std::function<void()> there, std::function<void()> back);
	public:
		template<typename O, typename R, std::enable_if_t<!std::is_same_v<R, void>, int> = 0, typename... Args>
		R add_commit_noeffect(StrongObject<O> obj, std::tuple<R> (O::*basefun)(StrongObject<Args>...), StrongObject<Args>... args) {
			Tag tag = get_tag();
			R val = std::get<0>(((*obj).*basefun)(args...));
			force_tag(tag);
			return val;
		}

		template<typename O, typename... Args>
		void add_commit_noeffect(StrongObject<O> obj, void (O::*basefun)(StrongObject<Args>...), StrongObject<Args>... args) {
			Tag tag = get_tag();
			((*obj).*basefun)(args...);
			force_tag(tag);
		}

		template<typename O, typename R, std::enable_if_t<!std::is_same_v<R, void>, int> = 0, typename... Args>
		void add_commit_dual(StrongObject<O> obj, R (O::*basefun)(StrongObject<Args>...), typename unpack_to_args<void, O, R>::type dualfun, StrongObject<Args>... args) {
			auto there_args = capture(args...);
			auto dual_args = std::apply([this](auto... args){return this->capture(args...);}, ((*obj).*basefun)(args...));
			auto there = [there_args{std::move(there_args)}, obj, basefun]() mutable {
				auto old_arg_tags = std::apply([](const auto&... a){return Journal::get_tags(a...);}, there_args);
				std::apply([obj, basefun](auto... a)mutable{((*obj).*basefun)(a...);}, captured_get(there_args));
				goto_tags(old_arg_tags, there_args);
			};
			auto back = [dual_args{std::move(dual_args)}, obj, dualfun]() mutable {
				auto old_arg_tags = std::apply([](const auto&... a){return Journal::get_tags(a...);}, dual_args);
				std::apply([obj, dualfun](auto... a)mutable{((*obj).*dualfun)(a...);}, captured_get(dual_args));
				goto_tags(old_arg_tags, dual_args);
			};
			add_new_vertex(std::move(there), std::move(back));
		}

		template<typename O, typename... Args>
		void add_commit_dual(StrongObject<O> obj, void (O::*basefun)(StrongObject<Args>...), void (O::*dualfun)(), StrongObject<Args>... args) {
			auto there_args = this->capture(args...);
			((*obj).*basefun)(args...);
			auto there = [there_args{std::move(there_args)}, obj, basefun]() mutable {
				auto old_arg_tags = std::apply([](const auto&... a){return Journal::get_tags(a...);}, there_args);
				std::apply([obj, basefun](auto... a)mutable{((*obj).*basefun)(a...);}, captured_get(there_args));
				goto_tags(old_arg_tags, there_args);
			};
			auto back = [obj, dualfun]() mutable {
				((*obj).*dualfun)();
			};
			add_new_vertex(std::move(there), std::move(back));
		}

		template<typename O, typename R, std::enable_if_t<!std::is_same_v<R, void>, int> = 0, typename... Args>
		std::tuple_element_t<0, R> add_commit_dual(StrongObject<O> obj, R (O::*basefun)(StrongObject<Args>...), typename unpack_to_args_skip_first<void, O, R>::type dualfun, StrongObject<Args>... args) {
			auto there_args = this->capture(args...);
			auto fun_ret = ((*obj).*basefun)(args...);
			std::tuple_element_t<0, R> return_value = std::move(std::get<0>(fun_ret));
			auto dual_args = std::apply([](auto... args){return capture(args...);}, trim_tuple(std::move(fun_ret)));
			auto there = [there_args{std::move(there_args)}, obj, basefun]() mutable {
				auto old_arg_tags = std::apply([](const auto&... a){return Journal::get_tags(a...);}, there_args);
				std::apply([obj, basefun](auto... a)mutable{((*obj).*basefun)(a...);}, captured_get(there_args));
				goto_tags(old_arg_tags, there_args);
			};
			auto back = [dual_args{std::move(dual_args)}, obj, dualfun]() mutable {
				auto old_arg_tags = std::apply([](const auto&... a){return Journal::get_tags(a...);}, dual_args);
				std::apply([obj, dualfun](auto... a)mutable{((*obj).*dualfun)(a...);}, captured_get(dual_args));
				goto_tags(old_arg_tags, dual_args);
			};
			add_new_vertex(std::move(there), std::move(back));
			return return_value;
		}

		template<typename O, typename R, std::enable_if_t<!std::is_same_v<R, void>, int> = 0, typename... Args>
		StrongObject<R> add_commit_dual(StrongObject<O> obj, std::tuple<StrongObject<R>> (O::*basefun)(StrongObject<Args>...), void (O::*dualfun)(), StrongObject<Args>... args) {
			auto there_args = this->capture(args...);
			auto fun_ret = ((*obj).*basefun)(args...);
			auto there = [there_args{std::move(there_args)}, obj, basefun]() mutable {
				auto old_arg_tags = std::apply([](const auto&... a){return Journal::get_tags(a...);}, there_args);
				std::apply([obj, basefun](auto... a)mutable{((*obj).*basefun)(a...);}, captured_get(there_args));
				goto_tags(old_arg_tags, there_args);
			};
			auto back = [obj, dualfun]() mutable {
				((*obj).*dualfun)();
			};
			add_new_vertex(std::move(there), std::move(back));
			return std::move(std::get<0>(fun_ret));
		}
	};
}

#include "overrides.h"

#endif
