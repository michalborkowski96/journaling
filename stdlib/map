import "container"
import "avl_"

nojournal MapComparator_ K V : Comparator<Pair<K, V>> {
	strong Comparator<K> c;
	constructor (Comparator<K> comp){
		c = comp;
	};
	fun bool less(Pair<K, V> e1, Pair<K, V> e2) {
		return c.less(e1.first, e2.first);
	};
}

Map K V : Container<Pair<K, V>> {
	strong Set<Pair<K, V>> data;
	constructor(Comparator<K> c) {
		data = new Set<Pair<K, V>>(new MapComparator_<K, V>(c));
	};
	fun V at(K key) {
		var Iterator<Pair<K, V>> iter = find(key);
		return iter.deref().second;
	} noeffect;
	fun int size() {
		return data.size();
	} noeffect;
	fun Iterator<Pair<K, V>> begin() {
		return data.begin();
	} noeffect;
	fun Iterator<Pair<K, V>> end() {
		return data.end();
	} noeffect;
	fun void clear() {
		data.clear();
	} irreversible;
	fun Iterator<Pair<K, V>> find(K key) {
		return data.find(new Pair<K, V>(key, null));
	} noeffect;

	fun MapEntry_<K, V> insert(K key, V value) {
		return this.insert(new Pair<K, V>(key, value));
	} dual (){};

	fun Pair<bool, Pair<K, V>> insert(Pair<K, V> entry) {
		var Pair<bool, Pair<K, V>> old = data.insert(entry);
		if(old.first) {
			return old, old.second, true;
		} else {
			return old, new Pair<K, V>(entry.first, null), false;
		}
	} dual (Pair<K, V> old, bool had_old) {
		if(had_old) {
			insert(old);
		} else {
			erase(old.first);
		}
	};

	fun Pair<bool, V> erase(K key) {
		var Pair<bool, Pair<K, V>> old = data.erase(entry);
		if(old.first) {
			return old, old.second, true;
		} else {
			return old, null, false;
		}
	} dual (Pair<K, V> old, bool had_old) {
		if(had_old) {
			insert(old);
		}
	};
}
