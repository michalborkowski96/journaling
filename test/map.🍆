import "../stdlib/better_string.🍆"
import "../stdlib/map.🍆"
import "../stdlib/vector.🍆"
import "../stdlib/journal_tag.🍆"

struct Main {
	strong OutputStream out;
	constructor(){
		out = new StdOut();
	};
	fun int main(){
		{
			auto ints = new UntaggedMap<Integer, BetterString>(new DefaultComparator<Integer>());
			ints.insert(new Integer(0), new BetterString("00"));
			ints.insert(new Integer(3), new BetterString("03"));
			ints.insert(new Integer(4), new BetterString("04"));
			ints.insert(new Integer(1), new BetterString("01"));
			ints.insert(new Integer(2), new BetterString("02"));
			ints.erase(new Integer(4));
			ints.print(out);
			out.print_newline();
			ints.find(new Integer(2)).deref().get_second().print(out);
			out.print_newline();
			auto t = new JournalTag<UntaggedMap<Integer, BetterString>>(ints);
			ints.erase(new Integer(2));
			ints.insert(new Integer(9), new BetterString("09"));
			ints.print(out);
			out.print_newline();
			ints.find(new Integer(9)).deref().get_second().print(out);
			out.print_newline();
			t.bring_back();
			ints.print(out);
			out.print_newline();
			ints.find(new Integer(2)).deref().get_second().print(out);
			out.print_newline();
		}

		out.print_newline();

		{
			auto ints = new Map<Integer, BetterString>(new DefaultComparator<Integer>());
			ints.insert(new Integer(0), new BetterString("00"));
			ints.insert(new Integer(3), new BetterString("03"));
			ints.insert(new Integer(4), new BetterString("04"));
			ints.insert(new Integer(1), new BetterString("01"));
			ints.insert(new Integer(2), new BetterString("02"));
			ints.erase(new Integer(4));
			ints.print(out);
			out.print_newline();
			ints.find(new Integer(2)).deref().get_second().print(out);
			out.print_newline();
			auto t = new JournalTag<Map<Integer, BetterString>>(ints);
			ints.erase(new Integer(2));
			ints.insert(new Integer(9), new BetterString("09"));
			ints.print(out);
			out.print_newline();
			ints.find(new Integer(9)).deref().get_second().print(out);
			out.print_newline();
			t.bring_back();
			ints.print(out);
			out.print_newline();
			ints.find(new Integer(2)).deref().get_second().print(out);
			out.print_newline();
		}

		out.print_newline();

		{
			auto strings = new Map<BetterString, BetterString>(new DefaultComparator<BetterString>());
			auto t = new JournalTag<Map<BetterString, BetterString>>(strings);
			var BetterString bs1 = new BetterString("bb");
			strings.insert(bs1, new BetterString("bs1"));
			strings.insert(new BetterString("cc"), new BetterString("cc"));
			strings.insert(new BetterString("dd"), new BetterString("dd"));
			strings.print(out);
			out.print_newline();
			bs1.set(0, 'e');
			bs1.set(1, 'e');
			strings.print(out);
			out.print_newline();
			strings.insert(new BetterString("aa"), new BetterString("aa"));
			strings.print(out);
			out.print_newline();
			strings.insert(new BetterString("ee"), new BetterString("ee"));
			strings.print(out);
			out.print_newline();
			bs1 = new BetterString("bb");
			strings.insert(bs1, new BetterString("new bs1"));
			strings.print(out);
			out.print_newline();
			bs1.set(0, 'e');
			bs1.set(1, 'e');
			strings.print(out);
			out.print_newline();
			strings.erase(new BetterString("bb"));
			strings.print(out);
			out.print_newline();
			strings.erase(new BetterString("ee"));
			strings.print(out);
			out.print_newline();
			auto tt = new JournalTag<Map<BetterString, BetterString>>(strings);
			t.bring_back();
			strings.print(out);
			out.print_newline();
			tt.bring_back();
			strings.print(out);
			out.print_newline();
		}
		return 0;
	};
}
