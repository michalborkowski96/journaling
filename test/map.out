{{0, 00}, {1, 01}, {2, 02}, {3, 03}}
02
{{0, 00}, {1, 01}, {3, 03}, {9, 09}}
09
{{0, 00}, {1, 01}, {2, 02}, {3, 03}}
02

{{0, 00}, {1, 01}, {2, 02}, {3, 03}}
02
{{0, 00}, {1, 01}, {3, 03}, {9, 09}}
09
{{0, 00}, {1, 01}, {2, 02}, {3, 03}}
02

{{bb, bs1}, {cc, cc}, {dd, dd}}
{{ee, bs1}, {cc, cc}, {dd, dd}}
{{aa, aa}, {ee, bs1}, {cc, cc}, {dd, dd}}
{{aa, aa}, {ee, bs1}, {cc, cc}, {dd, dd}, {ee, ee}}
{{aa, aa}, {bb, new bs1}, {cc, cc}, {dd, dd}, {ee, ee}}
{{aa, aa}, {ee, new bs1}, {cc, cc}, {dd, dd}, {ee, ee}}
{{aa, aa}, {cc, cc}, {dd, dd}, {ee, ee}}
{{aa, aa}, {cc, cc}, {dd, dd}}
{}
{{aa, aa}, {cc, cc}, {dd, dd}}
